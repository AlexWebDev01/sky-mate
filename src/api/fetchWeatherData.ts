import { Coordinates } from "../context/GlobalContext.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: string
) => {
  console.log("COORDINATES:", coordinates);

  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data = await response.json();

  return data;
};
