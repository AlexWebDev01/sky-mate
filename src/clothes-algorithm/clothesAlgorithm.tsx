import { TEMPERATURE_RANGES, CLOTHES_ADVICES } from "./clothesAlgorithmData";

const getTempRange = (temp: number): string => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  return "";
};

export const calculateClothesAdvice = (
  weatherCondition: string,
  temp: number
) => {
  const clothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return { adviceData: clothesAdvices[tempRangeName], tempRangeName };
  } else {
    console.log(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return null;
  }
};
