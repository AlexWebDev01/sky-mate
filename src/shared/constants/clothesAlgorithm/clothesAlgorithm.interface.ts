export enum WeatherConditions {
  rain = 'rain',
  snow = 'snow',
  sun = 'sun',
  clear = 'clear',
  clouds = 'clouds',
  fog = 'fog',
  dust = 'dust',
  mist = 'mist',
}

export enum TemperatureRangeNames {
  DANGEROUSLY_FREEZING = 'DANGEROUSLY_FREEZING',
  EXTREMELY_FREEZING = 'EXTREMELY_FREEZING',
  FREEZING = 'FREEZING',
  VERY_COLD = 'VERY_COLD',
  COLD = 'COLD',
  SLIGHTLY_COLD = 'SLIGHTLY_COLD',
  SLIGHTLY_FRESH = 'SLIGHTLY_FRESH',
  FRESH = 'FRESH',
  WARM = 'WARM',
  VERY_WARM = 'VERY_WARM',
  HOT = 'HOT',
  VERY_HOT = 'VERY_HOT',
  DANGEROUSLY_HOT = 'DANGEROUSLY_HOT',
}

export interface TempRanges {
  name: TemperatureRangeNames;
  lowestTemp: number;
  highestTemp: number;
}

export interface ClothesAdvice {
  clothesStyle: ClothesStyle;
  clothesList: string[];
  clothesDescription: string;
}

export interface ClothesAdviceWithTemp extends ClothesAdvice {
  tempRangeName: TemperatureRangeNames | null;
}

export type ClothesAdvices = {
  [K in TemperatureRangeNames]?: ClothesAdvice;
};

export interface ClothesAdvicesMap {
  rain: ClothesAdvices;
  sun: ClothesAdvices;
  clear: ClothesAdvices;
  clouds: ClothesAdvices;
  snow: ClothesAdvices;
  fog: ClothesAdvices;
  dust: ClothesAdvices;
  mist: ClothesAdvices;
}

export enum ClothesStyle {
  basic = 'Basic look',
  protect = 'Protect look',
  elegant = 'Elegant look',
  home = 'Home look',
  noLook = 'Oops, no look!',
}
