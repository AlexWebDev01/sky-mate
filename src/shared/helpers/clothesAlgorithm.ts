import {
  ClothesAdviceWithTemp,
  ClothesAdvices,
  TEMPERATURE_RANGE_NAMES,
  WeatherConditions,
  clothesStyle,
} from "../constants/clothesAlgorithm/clothesAlgorithm.interface";
import {
  TEMPERATURE_RANGES,
  CLOTHES_ADVICES,
} from "../constants/clothesAlgorithm/clothesAlgorithmData";

const getTempRange = (temp: number): TEMPERATURE_RANGE_NAMES => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  throw new Error(`There is no temperature range for provided temp: ${temp}`);
};

export const calculateClothesAdvice = (
  weatherCondition: WeatherConditions,
  temp: number
): ClothesAdviceWithTemp => {
  console.log(weatherCondition);
  const clothesAdvices: ClothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return {
      ...clothesAdvices[tempRangeName],
      tempRangeName,
    } as ClothesAdviceWithTemp;
  } else {
    console.error(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return {
      clothesStyle: clothesStyle.noLook,
      clothesList: [""],
      clothesDescription:
        "Ahh, the weather is so unpredictable, we don't have any advice for you! Try following your intuition or simply look at the sky.",
      tempRangeName: null,
    };
  }
};
