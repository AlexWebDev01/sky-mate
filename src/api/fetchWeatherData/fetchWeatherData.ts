import { Units, Coordinates, WeatherData } from "./fetchWeatherData.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: Units
): Promise<WeatherData> => {
  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data: WeatherData = await response.json();

  return data;
};
