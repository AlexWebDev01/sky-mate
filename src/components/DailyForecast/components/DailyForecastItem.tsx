import { DailyWeatherData } from "../../../api/fetchWeatherData/fetchWeatherData.interface";
import {
  calculateLocalDateAnotherFormat,
  calculateLocalFormattedDay,
} from "../../../shared/helpers/date/date";

interface Props {
  item: DailyWeatherData;
}

const DailyForecastItem = ({ item }: Props) => {
  return (
    <div className="daily-weather">
      <div className="day">{calculateLocalFormattedDay(item.dt)}</div>
      <div className={`${item.weather[0].main.toLowerCase()} date`}>
        {calculateLocalDateAnotherFormat(item.dt)}
      </div>
      <div className="day-temperature">{Math.round(item.temp.day)}&deg;</div>
      <div className="night-temperature">
        {Math.round(item.temp.night)}&deg;
      </div>
      <div className="weather">{item.weather[0].main}</div>
    </div>
  );
};

export default DailyForecastItem;
