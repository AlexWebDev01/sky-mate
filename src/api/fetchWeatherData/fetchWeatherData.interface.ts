import { WeatherConditions } from "../../shared/constants/clothesAlgorithm/clothesAlgorithm.interface";

export interface WeatherData {
  current: CurrentWeatherData;
  daily: DailyWeatherData[];
  hourly: HourlyWeatherData[];
  lat: number;
  lon: number;
  timezone: string;
  timezone_offset: number;
}

export interface CurrentWeatherData extends BaseWeatherData {
  sunrise: number;
  sunset: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface HourlyWeatherData extends BaseWeatherData {
  pop: number;
  wind_gust: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface DailyWeatherData extends BaseWeatherData {
  feels_like: { day: number; night: number; eve: number; morn: number };
  moon_phase: number;
  moonrise: number;
  moonset: number;
  pop: number;
  rain: number;
  summary: string;
  sunrise: number;
  sunset: number;
  temp: {
    day: number;
    eve: number;
    max: number;
    min: number;
    morn: number;
    night: number;
  };
  wind_gust: number;
}

interface BaseWeatherData {
  clouds: number;
  dew_point: number;
  dt: number;
  humidity: number;
  pressure: number;
  uvi: number;
  weather: Weather[];
  wind_deg: number;
  wind_speed: number;
}

export interface Weather {
  description: string;
  icon: string;
  id: number;
  main: WeatherConditions;
}

export interface Coordinates {
  lat: number;
  lon: number;
}

export enum Units {
  standard = "standard",
  metric = "metric",
  imperial = "imperial",
}
