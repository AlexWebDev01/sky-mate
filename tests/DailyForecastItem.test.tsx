import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import DailyForecastItem from '@components/DailyForecast/components/DailyForecastItem';
import { DailyWeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';
import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import {
  calculateLocalFormattedDay,
  calculateLocalDateAnotherFormat,
} from '@shared/helpers/date/date.ts';

jest.mock('@shared/helpers/date/date.ts', () => ({
  calculateLocalFormattedDay: jest.fn((dt) => `Day ${dt}`),
  calculateLocalDateAnotherFormat: jest.fn((dt) => `Formatted ${dt}`),
}));

const mockItem: DailyWeatherData = {
  clouds: 0,
  dew_point: 10,
  dt: 1625241600,
  humidity: 50,
  pressure: 1013,
  uvi: 5,
  weather: [
    {
      id: 800,
      main: WeatherConditions.clear,
      description: 'clear sky',
      icon: '01d',
    },
  ],
  wind_deg: 180,
  wind_speed: 5,
  feels_like: { day: 25, night: 15, eve: 20, morn: 10 },
  moon_phase: 0.5,
  moonrise: 1625205600,
  moonset: 1625258400,
  pop: 0.1,
  rain: 0,
  summary: 'Clear sky',
  sunrise: 1625212800,
  sunset: 1625263200,
  temp: {
    day: 25,
    eve: 20,
    max: 28,
    min: 15,
    morn: 18,
    night: 15,
  },
  wind_gust: 10,
};

describe('DailyForecastItem', () => {
  test('renders correctly with given item data', () => {
    const { getByText } = render(<DailyForecastItem item={mockItem} />);

    expect(getByText('Day 1625241600')).toBeInTheDocument();
    expect(getByText('Formatted 1625241600')).toBeInTheDocument();
    expect(getByText('25°')).toBeInTheDocument();
    expect(getByText('15°')).toBeInTheDocument();
    expect(getByText('clear')).toBeInTheDocument();
  });

  test('applies correct class based on weather type', () => {
    const { container } = render(<DailyForecastItem item={mockItem} />);
    const dateElement = container.querySelector('.clear.date');

    expect(dateElement).toBeInTheDocument();
  });

  test('calls utility functions with correct arguments', () => {
    render(<DailyForecastItem item={mockItem} />);

    expect(calculateLocalFormattedDay).toHaveBeenCalledWith(mockItem.dt);
    expect(calculateLocalDateAnotherFormat).toHaveBeenCalledWith(mockItem.dt);
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<DailyForecastItem item={mockItem} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
