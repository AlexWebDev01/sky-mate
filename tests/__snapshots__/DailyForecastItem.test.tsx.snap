// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyForecastItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="daily-weather"
    data-testid="daily-weather"
  >
    <div
      class="day"
    >
      Day 1625241600
    </div>
    <div
      class="clear date"
    >
      Formatted 1625241600
    </div>
    <div
      class="day-temperature"
    >
      25°
    </div>
    <div
      class="night-temperature"
    >
      15°
    </div>
    <div
      class="weather"
    >
      clear
    </div>
  </div>
</DocumentFragment>
`;
